### PR TITLE
Alias for logsumexp to special namespace

### PR DESCRIPTION
**Pull request by krshrimali**
See https://github.com/pytorch/pytorch/issues/50345

cc: @kshitij12345 @Lezcano @mruberry

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [inline aten/src/ATen/native/native_functions.yaml:9376]: Dispatch is not needed for alias. Plus explicitautograd is the correct key here.

File: torch/overrides.py
Comment (kshitij12345) [inline torch/overrides.py:536]: You dont need to remove this one.

File: torch/special/__init__.py
Comment (kshitij12345) [inline torch/special/__init__.py:162]: Update the documentaion of `torch.logsumexp` to point to `torch.special.logsumexp`

Eg. 
https://github.com/pytorch/pytorch/pull/55878/files#diff-547193dc33a84fcae0a5f7d9d4096c298f2bb718132e70a9cf5ec90bdfb113aaL4441

File: torch/testing/_internal/common_methods_invocations.py
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:6222]: `alias_remapping` test fails due to the way script is generated in `remapping_alias` vs `consistency_jit` tests.

```python
import torch

t = torch.zeros((2, 3, 3))
torch.special.logsumexp(t, (0,), False)


t = torch.zeros((2, 3, 3))
torch.special.logsumexp(t, [0, ], False)

variant = torch.special.logsumexp
t0 = t
s0 = (0,)
s1 = False

# As generated by `consistency_jit` test
# Works
script = """
def _fn(t0):
    return variant(t0, (0,), True)
"""

torch.jit.CompilationUnit(script)._fn

# As generated by `alias_remapping` test
# Doesn't work due to the type being generated as `tuple`.
script = """
def _fn(t0, s0: tuple, s1: bool):
    return variant(t0, s0, s1)
"""

torch.jit.CompilationUnit(script)._fn

```

File: aten/src/ATen/native/native_functions.yaml
Comment (krshrimali) [reply aten/src/ATen/native/native_functions.yaml]: Should be fixed now. Thanks!

File: torch/overrides.py
Comment (krshrimali) [reply torch/overrides.py]: Is there a reason why there are 2 same entries? I have added it back, but just asking to clarify. :)

File: torch/special/__init__.py
Comment (krshrimali) [reply torch/special/__init__.py:162]: Done, thanks for pointing this out, Kshiteej!

**Comment by krshrimali** [comment]
Updates:

Apologies for a long hold on this PR. Most of the issues have been resolved, waiting for the tests to pass then it will be ready for review. :)

File: torch/overrides.py
Comment (kshitij12345) [reply torch/overrides.py]: We are not actually removing `torch.logsumexp`. 

Also this test verifies if all the exposed function in the `torch` and few other namespaces have an entry for this test (so it would have failed mostly likely with this removed).

File: torch/csrc/api/include/torch/special.h
Comment (kshitij12345) [inline torch/csrc/api/include/torch/special.h:100]: We keep `Tensor& result` as the first argument to the function.

File: test/test_ops.py
Comment (kshitij12345) [inline test/test_ops.py:509]: Note for review:
This approach is consistent with `test_variant_consistency_jit`
Refer: https://github.com/pytorch/pytorch/pull/58838#discussion_r639805371

Will let other reviewers judge this part as I have limited knowledge about the same.

**Review by kshitij12345 — COMMENTED** [review]
Overall looks good to me. But one suggested change.

Thanks @krshrimali

File: torch/special/__init__.py
Comment (Lezcano) [inline torch/special/__init__.py:168]: We can take this chance to improve the wording of the documentation. What about just saying:

> Computes :math:\`\log \sum \exp (x)\` for an input `x` along the dimension :attr:\`dim\` with a numerically stable algorithm.

The math formula could be replaced even by `input.exp().sum(dim).log()`. What do you think @mruberry ?

File: torch/special/__init__.py
Comment (Lezcano) [inline torch/special/__init__.py:189]: This should be `torch.special.logsumexp`. 

It would also be more meaningful to write here:
> \>\>\> torch.dist(torch.special.logsumexp(a, 1), a.exp().sum(1).log())
> tensor(1.1921e-07)

**Review by lezcano — COMMENTED** [review]
Left a few small points. Overall looks very good, thanks @krshrimali

File: torch/csrc/api/include/torch/special.h
Comment (krshrimali) [reply torch/csrc/api/include/torch/special.h:100]: My bad, missed this change. Sorry!

File: aten/src/ATen/native/ReduceOps.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/ReduceOps.cpp:1018]: Good point! @mruberry had pointed the similar thing, initially. IIRC, there was an issue with supporting inplace arguments for operators unary operators like `erf`, etc. But for operators like `logsumexp` this does make sense and should be straightforward to do so. 

It would be nice to make `special_logsumexp` the primary implementation. (We can defer to doing it in a later PR as well)

Thanks for raising this point @Lezcano 

Details:
`special` variants support function and out variants. So to still support  inplace variant for lets say `erf`, we will have to use `special_erf_out` but that won't have a derivative formula so we will have to explicitly put an entry for `erf_` or use erf with copy for inplace variant.
Neither of them seemed convincing to make those changes. So we stuck to defer till the need arises for unary operators.

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:9517]: https://github.com/pytorch/pytorch/issues/50953

File: aten/src/ATen/native/ReduceOps.cpp
Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp:1018]: Sorry, but I did not understand. Why does `special_erf_out` not have a backwards formula?

Comment (kshitij12345) [reply aten/src/ATen/native/ReduceOps.cpp:1018]: `out` variants don't get the backwards formula.

File: test/test_ops.py
Comment (krshrimali) [reply test/test_ops.py]: Thanks for pointing this out, @Lezcano! Interestingly, we don't need this argument anymore since `args = ["to"]` and `args[1:]` is nothing but an empty list. I've removed this as an argument from method/inplace variants in the recent commit.

Comment (Lezcano) [reply test/test_ops.py]: Can you then update the comments accordingly?

File: torch/special/__init__.py
Comment (krshrimali) [reply torch/special/__init__.py:189]: I'm unfortunately on a different opinion with this change, I'm thinking that we should be consistent with examples on other functions and I feel having `>>> torch.special.logsumexp(a, 1)` is OK here.

Okay, I think your suggestion also makes sense, so to satisfy both simplicity as well as give a better meaning to what the function does. How about this instead?

```python
>>> torch.special.logsumexp(a, 1)
#output
>>> torch.dist(torch.special.logsumexp(a, 1), torch.log(torch.sum(torch.exp(a), dim)))
#output
```

File: torch/special/__init__.py
Comment (krshrimali) [reply torch/special/__init__.py:168]: I agree that this is better wording to the documentation, but do you think that this introduces redundancy (for formula) in the doc? As the formula is already mentioned just below this:

```
For summation index :math:`j` given by `dim` and other indices :math:`i`, the result is

    .. math::
        \text{{logsumexp}}(x)_{{i}} = \log \sum_j \exp(x_{{ij}})
```

I think for those who may not want to read the formula first, having it explained in words is OK. I maybe wrong but just sharing my thoughts. Looking forward to hear your opinion on this, @Lezcano @mruberry!

File: test/test_ops.py
Comment (krshrimali) [reply test/test_ops.py]: Since the `args=", ".join(args),` was removed but the comment still stays valid (as we do remove the first input tensor in `c` and `args_kw`), so I am not sure if there is any change in the comments needed. Please correct me if I'm missing anything here.

File: tools/autograd/gen_annotated_fn_args.py
Comment (krshrimali) [inline tools/autograd/gen_annotated_fn_args.py:38]: So far, added `special` functions had either just one argument (unary ufuncs) or expected `TensorLike()` args (`xlog1py`), so tests were passing (note: without this, the overrides test will test just `TensorLike()` inputs for the `nargs`), see:

https://github.com/pytorch/pytorch/blob/master/test/test_overrides.py#L600

```python
if func in annotated_args:
     # do stuff which special function should have done as well
     # Not True for special functions since args were not annotated yet
else:
    # ...
    nargs = len(args.args)
    if args.defaults is not None:
        nargs -= len(args.defaults)
    func_args = [instance_gen() for _ in range(nargs)]
```

Thanks to @kshitij12345 for suggesting the quick fix! :) This has been fixed for `fft` functions as well, to prevent future failures.

**Comment by krshrimali** [comment]
This is ready for review :) Tests failing seem irrelevant to this PR. PTAL whenever you find time, @mruberry @Lezcano @kshitij12345. Thanks for the patience.

File: test/test_ops.py
Comment (kshitij12345) [inline test/test_ops.py:480]: Maybe add a comment here

**Review by kshitij12345 — APPROVED** [review]
LGTM 

Thanks @krshrimali

File: test/test_ops.py
Comment (krshrimali) [reply test/test_ops.py:480]: Ah, right. Thanks for the suggestion. I'm thinking of something like: `# Required to avoid undefined value: tensor error in JIT compilation of the function template.`

Does this sound good to you? @kshitij12345

Comment (krshrimali) [reply test/test_ops.py:483]: Update: have added the comment.

**Comment by krshrimali** [comment]
Gentle ping: @mruberry - Can you please take a look whenever you find time? I understand you must be occupied for 1.9 release (🎉), but just thought I'll ping here in case this got lost in all the notifications :) Thanks!

File: aten/src/ATen/native/ReduceOps.cpp
Comment (krshrimali) [reply aten/src/ATen/native/ReduceOps.cpp:1018]: Thanks for the discussion, @kshitij12345 and @Lezcano. Since this PR also blocks `polygamma` PR (because of changes in `alias_remapping` test), I'll like to make the `special_logsumexp` implementation as the primary one in another PR.

File: torch/special/__init__.py
Comment (mruberry) [reply torch/special/__init__.py:203]: Let's not sweat the documentation in this pr

**Comment by mruberry** [comment]
> Gentle ping: @mruberry - Can you please take a look whenever you find time? I understand you must be occupied for 1.9 release (🎉), but just thought I'll ping here in case this got lost in all the notifications :) Thanks!

Thanks for the ping.

What's going on with the JIT test changes in this PR? Are they required, and what do they fix?

**Comment by mruberry** [comment]
I would leave the docs in torch for now and have the alias go from special to torch. Eventually we'll need to reconcile which one claims to be the alias, but we'll need to develop some principled way of doing so. Possibly based on usage data or Python Array API or NumPy compatibility.

**Comment by krshrimali** [comment]
> What's going on with the JIT test changes in this PR? Are they required, and what do they fix?

The `test_jit_alias_remapping` was failing earlier because `tuple` input was passed to the function which it currently doesn't support in the function template (it expects a list as an input). 

**Earlier**:

https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L548-L551
```python
fn_template = '''
    def _fn({args_annot_kw}):
        return variant({args_kw})
'''
```

where `args_kw` will look like `s0: int, s1: tuple` (just an example) and thus the the error (with tuple in the template). Here is where `args_kw` is defined:

https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L510-L512
```python
args_kw = args + \
    [f"s{i}" for i in range(len(sample.args))] + \
    [f"{k}={quote_strs(v)}" for k, v in sample.kwargs.items()]
```

The solution I'm proposing here is to _not_ have the types annotated in the function template and instead pass the inputs directly.

**After**:

`args_kw` has been changed to take the value instead of adding a placeholder (I just realized that this defers from the name of the variable: `args_kw`):

```python

**Comment by kshitij12345** [comment]
To add a little more context. This is how we acquire types for annotation currently (using `type(obj).__name__`).
https://github.com/pytorch/pytorch/blob/3dc8112187c5a4162581b9725695455ca959e752/test/test_ops.py#L839-L841
which works correctly most of the times. But causes problems with Sequence types.

I think the new fix is correct as it follows the similar approach to `test_variant_consisteny_jit` generates the scripted and the traced functions.

File: test/test_ops.py
Comment (kshitij12345) [inline test/test_ops.py:829]: I think it would be nice to have an example of what is happening (similar to what was present)

Comment (krshrimali) [reply test/test_ops.py]: Done, thanks!

**Comment by krshrimali** [comment]
Gentle ping, @mruberry - Can you please take a look at this whenever you find time? Thanks!

File: torch/special/__init__.py
Comment (Lezcano) [reply torch/special/__init__.py]: I think having both would be a good adition. Note that the formula I wrote had an errata, and  `dim` should be `1`.

**Review by lezcano — COMMENTED** [review]
I do not know enough to be able to say whether the changes in the JIT are fine, I'll let @mruberry assess that or find the person that knows about this. Everything else looks good to me. Thank you Kush!

On this topic, I have a question. Would this fix happen to fix all (or a number of) those tests that we used to ignore in the OpInfos, or does it address something different? @krshrimali @kshitij12345 
If it is so, this is great! We should then remove the relevan skips.
If it wasn't, how do other functions that have this problem deal with the JIT errors that we found here?

**Comment by krshrimali** [comment]
> I do not know enough to be able to say whether the changes in the JIT are fine, I'll let @mruberry assess that or find the person that knows about this. Everything else looks good to me. Thank you Kush!

Thanks, @Lezcano for taking a closer look at this PR. Appreciate your comments. Will update the doc to add both formulae.

> On this topic, I have a question. Would this fix happen to fix all (or a number of) those tests that we used to ignore in the OpInfos, or does it address something different? @krshrimali @kshitij12345

This is a great question, the fix in this PR is for `test_jit_alias_remapping` (that is only for JIT remapping tests on aliases). I just noticed that there are a few tests which are skipped, and which _probably_ can be solved by the fix from this PR. I'll see if they pass on this branch.

> If it is so, this is great! We should then remove the relevan skips.

Agreed! :)

**Comment by krshrimali** [comment]
**Update:**

There are currently 2 tests skipped:

1. `vstack` which is because of the function taking sequence of tensors as input instead of just a tensor. This is not fixed in this PR but again we can remove `i0` from the formals and directly pass sample inputs to actuals, but I am not too sure if this is the _correct_ solution for this issue.
2. `movedim` - this has been fixed by this PR. :)

Thanks for asking the right question, @Lezcano!

cc: @mruberry @Lezcano @kshitij12345

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:4989]: Let's keep the docs in torch for now and have special.logsumexp alias it.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7117]: Awesome skip removal

**Review by mruberry — COMMENTED** [review]
This looks really good. cc @eellison, who's interested in reducing jit test skips.

@eellison, do these jit test changes make sense?

I also made one comment about keeping the docs on torch.logsumexp for now (we'll review what goes in special vs. torch later)

**Comment by krshrimali** [comment]
Hi, @eellison and @mruberry - Gentle ping if this can be looked at. Thanks!

**Review by eellison — APPROVED** [review]
Thanks for removing the skipped test 😍 😍 😍  LGTM if tests pass

**Review by mruberry — APPROVED** [review]
LGTM! Thanks @krshrimali

**Pull request closed without merging**
